server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamRouteProvider.java

```diff
@@ -30,90 +27,88 @@ public class UpstreamRouteProvider {
+            }
+            result.setLastAccessTime(System.currentTimeMillis());
+            return result;
+        });
+        return new UpstreamRoute(balancer, MAX_RETRY_COUNT);
+    }
 
-            RouteEndpointProvider endpointProvider = new RouteEndpointProvider(route);
-            TieredBalancer previous = oldState.get(name);
-            updateDeployment(endpointProvider, previous, newState);
+    private List<Upstream> getUpstreams(Deployment deployment) {
+        if (deployment instanceof Model model && !model.getUpstreams().isEmpty()) {
+            return model.getUpstreams();
         }
-    }
 
-    private static void updateDeployments(Map<String, TieredBalancer> newState, Map<String, TieredBalancer> oldState,
-                                          Collection<? extends Deployment> deployments) {
-        for (Deployment deployment : deployments) {
-            String name = deployment.getName();
+        Upstream upstream = new Upstream();
+        upstream.setEndpoint(deployment.getEndpoint());
+        upstream.setKey("whatever");
+        return List.of(upstream);
+    }
 
-            DeploymentUpstreamProvider endpointProvider = new DeploymentUpstreamProvider(deployment);
-            TieredBalancer previous = oldState.get(name);
-            updateDeployment(endpointProvider, previous, newState);
-        }
+    private String getKey(Route route) {
+        Objects.requireNonNull(route);
+        return "route:" + route.getName();
     }
 
-    private static void updateDeployment(UpstreamProvider upstream, TieredBalancer previous, Map<String, TieredBalancer> newState) {
-        String name = upstream.getName();
-        TieredBalancer balancer;
-        if (previous != null && isUpstreamsTheSame(upstream, previous)) {
-            balancer = previous;
+    private String getKey(Deployment deployment) {
+        Objects.requireNonNull(deployment);
+        String prefix;
+        if (deployment instanceof Model) {
+            prefix = "model";
+        } else if (deployment instanceof Application) {
+            prefix = "application";
+        } else if (deployment instanceof Assistant) {
+            prefix = "assistant";
         } else {
-            balancer = new TieredBalancer(name, upstream.getUpstreams());
+            throw new IllegalArgumentException("Unsupported deployment type: " + deployment.getClass().getName());
         }
-        TieredBalancer previousBalancer = newState.putIfAbsent(name, balancer);
-        if (previousBalancer != null) {
-            log.warn("Duplicate deployment name: {}", name);
+        return prefix + ":" + deployment.getName();
+    }
+
+    private void evictExpiredBalancers() {
+        long currentTime = System.currentTimeMillis();
+        for (Map.Entry<String, TieredBalancer> entry : balancers.entrySet()) {
+            TieredBalancer balancer = entry.getValue();
+            if (currentTime - balancer.getLastAccessTime() > IDLE_PERIOD_IN_MS) {
```

Comment:
should be updated with the same memory-semantic as you set value. So map.compute and check there.

Reply:
fixed